### PR TITLE
helmfile: set HELM_PLUGINS if pluginsDir is given

### PR DESCRIPTION
**Pull request by yurrriq**
###### Description of changes

RFC on a way to address #217768. /cc @sleiner @06kellyjac

TODO:
- [x] update release notes.
 
###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [x] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/networking/cluster/helmfile/default.nix [hunk 1]
```diff
@@ -19,9 +25,14 @@ buildGoModule rec {
 
   ldflags = [ "-s" "-w" "-X go.szostok.io/version.version=v${version}" ];
 
-  nativeBuildInputs = [ installShellFiles ];
+  nativeBuildInputs =
+    [ installShellFiles ] ++
+    lib.optionalString (pluginsDir != null) [ makeWrapper ];
```
Comment (Staberinde) [inline pkgs/applications/networking/cluster/helmfile/default.nix:30]: I don't think `optionalString` will work here, you might need to use the full `if ... then ... else` syntax

**Comment by Staberinde** [comment]
I've tested this out locally via patching, I think there's one more piece missing to get this working out-of-the-box; something along the lines of:
```diff
diff --git a/pkgs/top-level/all-packages.nix b/pkgs/top-level/all-packages.nix
index 5a46b6afe7f..25cc8732c54 100644
--- a/pkgs/top-level/all-packages.nix
+++ b/pkgs/top-level/all-packages.nix
@@ -38619,7 +38619,9 @@ with pkgs;

   helm = callPackage ../applications/audio/helm { };

-  helmfile = callPackage ../applications/networking/cluster/helmfile { };
+  helmfile = callPackage ../applications/networking/cluster/helmfile {
+    pluginsDir = kubernetes-helm.passthru.pluginsDir;
+  };

   helmsman = callPackage ../applications/networking/cluster/helmsman { };
```

File: pkgs/applications/networking/cluster/helmfile/default.nix (hunk 1 above)
Comment (yurrriq) [reply pkgs/applications/networking/cluster/helmfile/default.nix:30]: Yep... I meant `lib.optionals`. Thanks.

**Comment by yurrriq** [comment]
@Staberinde, thanks for the feedback. Re: out-of-the-box functionality. Currently, `kubernetes-helm` doesn't wrap any plugins and doesn't expose `passthru.pluginsDir`. I think that's how it should be. But perhaps we should add a top-level `helmfile-wrapped` like this.
```nix
{
  helmfile-wrapped = callPackage ../applications/networking/cluster/helmfile {
    pluginsDir = kubernetes-helm-wrapped.passthru.pluginsDir;
  };
}
```

But even `kubernetes-helm-wrapped` doesn't come with any plugins by default...

My thinking was that users would wrap their own helm derivation and pass that to helmfile. Something like this.
```nix
final: prev:
rec {
  myHelm = prev.wrapHelm prev.kubernetes-helm {
    plugins = with prev.kubernetes-helmPlugins; [
      helm-diff
      helm-secrets
    ];
  };
  myHelmfile = prev.helmfile.override {
    inherit (myHelm.passthru) pluginsDir;
  };
}
```

What do you think?

**Comment by yurrriq** [comment]
I've added some release notes. I'd still like some feedback, but I think this is ready to merge (and backport to 23.05).

**Comment by yurrriq** [comment]
Result of `nixpkgs-review pr 231312` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>helmfile-wrapped</li>
  </ul>
</details>

**Comment by yurrriq** [comment]
I've rebased and since it's been a while and we got a couple of :tada: reactions, I guess no one's opposed, so I'll go ahead and merge.

**Pull request merged**